packages/http-crawler/src/internals/http-crawler.ts

```diff
@@ -0,0 +1,845 @@
+            }
+
+            throw e;
+        }
+    }
+
+    /**
+     * Encodes and parses response according to the provided content type
+     */
+    protected async _parseResponse(request: Request, responseStream: IncomingMessage, crawlingContext: Context) {
+        const { statusCode } = responseStream;
+        const { type, charset } = parseContentTypeFromResponse(responseStream);
+        const { response, encoding } = this._encodeResponse(request, responseStream, charset);
+        const contentType = { type, encoding };
+
+        if (statusCode! >= 500) {
+            const body = await readStreamToString(response, encoding);
+
+            // Errors are often sent as JSON, so attempt to parse them,
+            // despite Accept header being set to text/html.
+            if (type === APPLICATION_JSON_MIME_TYPE) {
+                const errorResponse = JSON.parse(body);
+                let { message } = errorResponse;
+                if (!message) message = util.inspect(errorResponse, { depth: 1, maxArrayLength: 10 });
+                throw new Error(`${statusCode} - ${message}`);
+            }
+
+            // It's not a JSON, so it's probably some text. Get the first 100 chars of it.
+            throw new Error(`${statusCode} - Internal Server Error: ${body.slice(0, 100)}`);
+        } else if (HTML_AND_XML_MIME_TYPES.includes(type)) {
+            const isXml = type.includes('xml');
+            const parsed = await this._parseHTML(response, isXml, crawlingContext);
+            return { ...parsed, isXml, response, contentType };
+        } else {
+            const body = await concatStreamToBuffer(response);
+            return { body, response, contentType };
+        }
+    }
+
+    protected async _parseHTML(response: IncomingMessage, _isXml: boolean, _crawlingContext: Context): Promise<Partial<Context>> {
+        return {
+            body: await concatStreamToBuffer(response),
+        } as Partial<Context>;
+    }
+
+    /**
+     * Combines the provided `requestOptions` with mandatory (non-overridable) values.
+     */
+    protected _getRequestOptions(request: Request, session?: Session, proxyUrl?: string, gotOptions?: OptionsInit) {
+        const requestOptions: OptionsInit & { isStream: true } = {
+            url: request.url,
+            method: request.method as Method,
+            proxyUrl,
+            timeout: { request: this.navigationTimeoutMillis },
+            sessionToken: session,
+            ...gotOptions,
+            headers: { ...request.headers, ...gotOptions?.headers },
+            https: {
+                ...gotOptions?.https,
+                rejectUnauthorized: !this.ignoreSslErrors,
+            },
+            isStream: true,
+        };
+
+        // Delete any possible lowercased header for cookie as they are merged in _applyCookies under the uppercase Cookie header
+        Reflect.deleteProperty(requestOptions.headers!, 'cookie');
```

Comment:
Does this do the same thing as the previous code?

Reply:
If by the previous code you mean the current cheeriocrawler, then yes. I haven't modified any logic except extracted the cheerio thing to another class

Reply:
yeah this exact line is in master too, just elsewhere, lets keep this as close to the original as possible so we dont introduce some hidden BCs